### ISSUE TITLE
Very slow compile on simple model

### ISSUE THREAD

**Issue by jhmenke** [pyro-ppl/numpyro#543]
Below is an example code of using MCMC to infer some variables *vm* (size: 3) and *va* (size: 2). However, on my machine the runtime of the code is about 3 minutes, mostly due to an abnormally slow compilation process. The results are exactly what i would expect though.

Unfortunately, debugging that is not really my expertise, therefore i have created a minimal example and also attached the compile logs as per the suggestion. I think there is a good chance this is some beginners mistake and not actually a bug in your code, but i'm not quite sure to be honest.

I'm running an Ubuntu derivate 64 bit with Python 3.7 and updated jax/jaxlib/numpyro releases. Thanks to someone who can take a quick peek at this!

> ********************************
> Very slow compile?  If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
> ********************************

```
import argparse

import jax
import jax.numpy as np
import jax.random as random
import numpy as onp
import numpyro
import numpyro.distributions as dist
import pandas as pd
from numpyro.distributions.constraints import interval, greater_than
from numpyro.infer import MCMC, NUTS

jax.config.update('jax_platform_name', 'cpu')

# create the data to use
Z = onp.asarray([[onp.nan, 1e-2 + 3e-2j, 2e-2 + 5e-2j],
                 [1e-2 + 3e-2j, onp.nan, 3e-2 + 8e-2j],
                 [2e-2 + 5e-2j, 3e-2 + 8e-2j, onp.nan]], dtype=onp.complex)
Y = -onp.reciprocal(Z)
Y[onp.where(onp.isnan(Y))] = 0.
g, b = np.array(-Y.real), np.array(-Y.imag)
Y[0, 0] = -Y[0, :].sum()
Y[1, 1] = -Y[1, :].sum()
Y[2, 2] = -Y[2, :].sum()
G, B = np.array(Y.real), np.array(Y.imag)
measurements_df = pd.DataFrame(columns=["type", "value", "std"],
                               index=pd.RangeIndex(8),
                               data=[["pl_1_2", 0.888, 0.008],
                                     ["pl_1_3", 1.173, 0.008],
                                     ["p_2", -0.501, 0.01],
                                     ["ql_1_2", 0.568, 0.008],
                                     ["ql_1_3", 0.663, 0.008],
                                     ["q_2", -0.286, 0.01],
                                     ["v_1", 1.006, 0.004],
                                     ["v_2", 0.968, 0.004]])

# define equations
def p_i(i: int, v, delta):
    res = 0.
    for j in range(len(v)):
        res = np.add(res, v[j] * (G[i, j] * np.cos(delta[i] - delta[j]) + B[i, j] * np.sin(delta[i] - delta[j])))
    return v[i] * res


def q_i(i: int, v, delta):
    res = 0.
    for j in range(len(v)):
        res = np.add(res, v[j] * (G[i, j] * np.sin(delta[i] - delta[j]) - B[i, j] * np.cos(delta[i] - delta[j])))
    return v[i] * res


def p_ij(i: int, j: int, v, delta, g, b):
    return v[i] * v[i] * (g[i, j]) - v[i] * v[j] * (g[i, j] * np.cos(delta[i] - delta[j]) + b[i, j] * np.sin(delta[i] - delta[j]))


def q_ij(i: int, j: int, v, delta, g, b):
    return -v[i] * v[i] * (b[i, j]) - v[i] * v[j] * (g[i, j] * np.sin(delta[i] - delta[j]) - b[i, j] * np.cos(delta[i] - delta[j]))


# define model
def se_model(measurements: np.array, stds: np.array, n_buses: int):
    vm_dist = dist.Normal(np.ones(n_buses), 0.03 * np.ones(n_buses))
    vm_dist.arg_constraints = {
        "loc": interval(0.9, 1.1),
        "scale": greater_than(0)
    }
    vm = numpyro.sample("vm", vm_dist)
    va_dist = numpyro.sample("va", dist.Uniform(-0.1 * np.ones(n_buses - 1), 0.1 * np.ones(n_buses - 1)))
    va = np.concatenate([np.zeros(1), va_dist])
    pl_0_1 = p_ij(0, 1, vm, va, g, b)
    pl_0_2 = p_ij(0, 2, vm, va, g, b)
    p_1 = p_i(1, vm, va)
    ql_0_1 = q_ij(0, 1, vm, va, g, b)
    ql_0_2 = q_ij(0, 2, vm, va, g, b)
    q_1 = q_i(1, vm, va)
    all_meas_values = np.stack([pl_0_1, pl_0_2, p_1, ql_0_1, ql_0_2, q_1, vm[0], vm[1]])
    numpyro.sample("Y", dist.Normal(all_meas_values, stds), obs=measurements)


def run_inference(model, args, rng_key, measurements):
    kernel = NUTS(model)
    mcmc = MCMC(kernel, args.num_warmup, args.num_samples, num_chains=args.num_chains, progress_bar=True)
    mcmc.run(rng_key, np.array(measurements["value"].values), np.array(measurements["std"].values), 3)
    mcmc.print_summary()
    return mcmc.get_samples()


def main(args):
    rng_key, rng_key_predict = random.split(random.PRNGKey(0))
    samples = run_inference(se_model, args, rng_key, measurements_df)
    print(samples)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Simple Bayesian state estimation")
    parser.add_argument("-n", "--num-samples", nargs="?", default=700, type=int)
    parser.add_argument("--num-warmup", nargs='?', default=300, type=int)
    parser.add_argument("--num-chains", nargs='?', default=1, type=int)
    parser.add_argument("--device", default='cpu', type=str, help='use "cpu" or "gpu".')
    args = parser.parse_args()
    numpyro.set_platform(args.device)
    numpyro.set_host_device_count(args.num_chains)
    main(args)

```

[xla_compile_logs.zip](https://github.com/pyro-ppl/numpyro/files/4245158/xla_compile_logs.zip)

**Comment by fehiepsi** [comment]
This is unfortunate. I'm not sure what we can do to improve JAX compiling time... You can temporarily fix the issue by removing [this line](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/mcmc.py#L254). We will investigate more about it.

**Comment by fehiepsi** [comment]
@neerajprad In #127, we added an arg `heuristic_step_size` to disable the functionality of `find_reasonable_step_size`. In preparing for 0.2 release, we [remove it](https://github.com/pyro-ppl/numpyro/pull/310) because using `lax.scan` solves the slowness of the time series forecasting tutorial or the hmm example IIRC (some discussion [here](https://github.com/pyro-ppl/numpyro/issues/144#issuecomment-491468152)). Should we add it back while investigating more about why the slowness happens with the `find_reasonable_step_size` function?

**Comment by jhmenke** [comment]
> This is unfortunate. I'm not sure what we can do to improve JAX compiling time... You can temporarily fix the issue by removing [this line](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/mcmc.py#L254). We will investigate more about it.

Thank you so much! It is now down to 11s from 3 minutes.

**Comment by jhmenke** [comment]
@fehiepsi You are not referring to this call, are you?

https://github.com/pyro-ppl/numpyro/blob/f2aefcf1b034418244c1e465c64bce469388e566/numpyro/infer/hmc_util.py#L383

That finishes in ~3 seconds in debug mode.

**Comment by fehiepsi** [comment]
I think you would need to increase num_warmup. 300 might be not enough for this model.

**Comment by jhmenke** [comment]
Thanks! Unfortunately, even using 100 000 warmup samples makes it converge to the wrong value.

First plot is without any code changes and the low number of samples:
![bokeh_plot](https://user-images.githubusercontent.com/25080218/75322436-05340500-5873-11ea-9f8b-e0bfa8e5dcbe.png)

This is with the single line removed and 100k warmup samples and 100k samples

![bokeh_plot (1)](https://user-images.githubusercontent.com/25080218/75322473-18df6b80-5873-11ea-928b-e501a77adc41.png)

**Comment by fehiepsi** [comment]
@jhmenke I think that disabling `find_reasonable_step_size` does not cause the issue. Probably you are facing [identifiability](https://mc-stan.org/docs/2_22/stan-users-guide/collinearity-section.html) issue. Could you verify that the inference gives consistent results with different random seeds? Another way is to verify your posterior predictive.

**Comment by jhmenke** [comment]
I can confirm that using different random seeds produces different results, i.e., it's not an issue with numpyro.

**Comment by fehiepsi** [comment]
Just update: the `find_reasonable_step_size` functionality is disabled by default.
